pkg/transport/fetchidentity.go

```diff
@@ -0,0 +1,101 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package transport
+
+import (
+	"context"
+	"net"
+
+	"github.com/zeebo/errs"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"storj.io/storj/pkg/identity"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/peertls"
+)
+
+// TODO: check whether this would be problematic with concurrency or should we thread the authinfo through context?
+type handshakeCapture struct {
+	credentials.TransportCredentials
+	AuthInfo credentials.AuthInfo
+}
+
+// ClientHandshake does the authentication handshake specified by the corresponding
+// authentication protocol on conn for clients. It returns the authenticated
+// connection and the corresponding auth information about the connection.
+func (capture *handshakeCapture) ClientHandshake(ctx context.Context, s string, conn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	conn, auth, err := capture.TransportCredentials.ClientHandshake(ctx, s, conn)
+	capture.AuthInfo = auth
+	return conn, auth, err
+}
+
+// ServerHandshake does the authentication handshake for servers. It returns
+// the authenticated connection and the corresponding auth information about
+// the connection.
+func (capture *handshakeCapture) ServerHandshake(conn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	conn, auth, err := capture.TransportCredentials.ServerHandshake(conn)
+	capture.AuthInfo = auth
+	return conn, auth, err
+}
+
+// FetchPeerIdentity dials the node and fetches the identity
+func (transport *Transport) FetchPeerIdentity(ctx context.Context, node *pb.Node, opts ...grpc.DialOption) (_ *identity.PeerIdentity, err error) {
+	defer mon.Task()(&ctx, "node: "+node.Id.String()[0:8])(&err)
+
+	if node.Address == nil || node.Address.Address == "" {
+		return nil, Error.New("no address")
+	}
+	tlsConfig := transport.tlsOpts.ClientTLSConfig(node.Id)
+
+	capture := &handshakeCapture{
+		TransportCredentials: credentials.NewTLS(tlsConfig),
+	}
+
+	options := append([]grpc.DialOption{
+		grpc.WithTransportCredentials(capture),
+		grpc.WithBlock(),
+		grpc.FailOnNonTempDialError(true),
+		grpc.WithContextDialer(func(ctx context.Context, addr string) (net.Conn, error) {
+			conn, err := (&net.Dialer{}).DialContext(ctx, "tcp", addr)
+			if err != nil {
+				return nil, err
```

Comment:
nit - error wrap?

Reply:
no, because we are here, no, because it is wrapping internal behavior of grpc and we don't know how the error is handled internally, by wrapping we might lose important interface methods.

Reply:
this one bubbles up in the call `conn, err := grpc.DialContext(timedCtx, node.GetAddress().Address, options...)`, no?

Reply:
Not sure what you mean?

Reply:
When this function (which is an option) is called and due that its result is returned?

I think that the function gets called by `grpc.DialContext(...` as its receiving the options.